job/jobcache.go

```diff
@@ -1,36 +1,88 @@
 package job
 
 import (
+	"os"
+	"os/signal"
 	"sync"
+	"time"
 )
 
-var (
-	AllJobs = &JobCache{
-		Jobs:   make(map[string]*Job),
-		rwLock: sync.Mutex{},
+type JobCache interface {
+	Get(id string) *Job
+	GetAll() map[string]*Job
+	Set(j *Job)
+	Delete(id string)
+	Persist() error
+}
+
+type MemoryJobCache struct {
+	// Jobs is a map from Job id's to pointers to the jobs.
+	// Used as the main "data store" within this cache implementation.
+	jobs            map[string]*Job
+	rwLock          sync.Mutex
+	jobDB           JobDB
+	persistWaitTime time.Duration
+}
+
+func NewMemoryJobCache(jobDB JobDB, persistWaitTime time.Duration) *MemoryJobCache {
+	if persistWaitTime == 0 {
+		persistWaitTime = 5 * time.Second
 	}
-)
+	return &MemoryJobCache{
+		jobs:            map[string]*Job{},
+		rwLock:          sync.Mutex{},
+		jobDB:           jobDB,
+		persistWaitTime: persistWaitTime,
+	}
+}
+
+func (c *MemoryJobCache) Start() {
+	// Prep cache
+	allJobs, err := c.jobDB.GetAll()
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, j := range allJobs {
+		j.StartWaiting(c)
+		c.Set(j)
+	}
+
+	// Occasionally, save items in cache to db.
+	go c.PersistEvery()
+
+	// Process-level defer for shutting down the db.
+	ch := make(chan os.Signal)
+	signal.Notify(ch, os.Interrupt, os.Kill)
+	go func() {
+		s := <-ch
+		log.Info("Process got signal: %s", s)
+		log.Info("Shutting down....")
 
-type JobCache struct {
-	Jobs   map[string]*Job
-	rwLock sync.Mutex
+		// Persist all jobs to database
+		c.Persist()
+
+		// Close the database
+		c.jobDB.Close()
+
+		os.Exit(1)
+	}()
 }
 
-func (c *JobCache) Get(id string) *Job {
+func (c *MemoryJobCache) Get(id string) *Job {
 	c.rwLock.Lock()
 	defer c.rwLock.Unlock()
 
-	return c.Jobs[id]
+	return c.jobs[id]
 }
 
-func (c *JobCache) GetAll() map[string]*Job {
+func (c *MemoryJobCache) GetAll() map[string]*Job {
 	c.rwLock.Lock()
 	defer c.rwLock.Unlock()
 
-	return c.Jobs
+	return c.jobs
```

Comment:
@ajvb This doesn't solve it, because the called still has a reference to this `map` and can do whatever it wants with it, including read from it or change it without any locking occurring.